quil-plotting/quil_plotting/dataframe.py

```diff
@@ -0,0 +1,346 @@
+    assert df[color_by].dtype == "category", f"{color_by} is not a categorical field. Options include: {categoricals}"
+    if normalize_by is not None:
+        assert (
+            df[normalize_by].dtype == "category"
+        ), f"{normalize_by} is not a categorical field. Options include: {categoricals}"
+    if label_by is not None:
+        assert (
+            df[label_by].dtype == "category"
+        ), f"{label_by} is not a categorical field. Options include: {categoricals}"
+
+    # Determine the order of the runners and calculate the y-offset for the IQ values.
+    if runner_order in df.columns:
+        sorted_df = df.sort_values(["Time (s)"])
+        runner_values = sorted_df[runners].drop_duplicates()
+    elif runner_order == "Topological":
+        # need the root instructions
+        raise NotImplementedError()
+    else:
+        runner_values = df[runners].drop_duplicates()
+    runner_map = {runner: idx for idx, runner in enumerate(runner_values)}
+    df["Offset"] = df[runners].apply(lambda runner: runner_map.get(runner, 0))
+
+    # Calculate colors
+    color_cycle = cycle(Color)
+    color_values = df[color_by].cat.categories
+    # pre-defined colors
+    color_map = {"ISWAP": Color.YELLOW, "RX": Color.GREEN, "CZ": Color.BLUE}
+    # add additional colors as needed
+    color_map |= {val: next(color_cycle) for val in color_values if val not in color_map}
+    df["Color"] = df[color_by].apply(lambda x: hex_to_rgba(color_map.get(x, "#ffffff")))
+    df["Color Faded"] = df[color_by].apply(lambda x: hex_to_rgba(color_map.get(x, "#ffffff"), 0.3))
+
+    # Normalize
+    if normalize_by is None:
+        df["Normalized IQ"] = df["IQ"]
+    else:
+        df["Magnitude"] = np.abs(df["IQ"])
+        normalization_values = df.groupby(normalize_by, observed=True)["Magnitude"].transform("max")
+        df["Normalized IQ"] = 0.45 * df["IQ"] / normalization_values
+
+    # Label
+    if label_by is None:
+        label_by = runners
+
+    df["Label"] = df[label_by]
+    df["Runner"] = df[runners]
+
+    df = df.astype(
+        {
+            "Instruction Index": int,
+            "Instruction": "category",
+            "Logical Instruction": "category",
+            "Gate": "category",
+            "Qubit": "category",
+            "Pulse": "category",
+            "Waveform": "category",
+            "Start Time (s)": float,
+            "End Time (s)": float,
+            "Duration (s)": float,
+            "Pulse Time (s)": float,
+            "Frame": "category",
+            "Channel Type": "category",
+            "IQ": complex,
+            "Offset": float,
+            "Color": "category",
+            "Color Faded": "category",
+            "Normalized IQ": complex,
+            "Label": "category",
+            "Runner": "category",
+        }
+    )
+
+    return df
+
+
+def deoxidize(num) -> complex:
+    """Remove Rust from a Number or PrefixExpression."""
```

Comment:
This mentions the `Number` variant, but doesn't handle it. Also, the `evaluate` method is available on any `Expression`, and raises an error if it cannot evaluate the expression to a complex number. So this could be simplified to a single line while keeping the same behavior. Since it is a single line, this function can likely be removed.

```python
num.evaluate({}, {})
```

Reply:
This doesn't seem to work on PrefixExpressions. Will revisit it though.